core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -191,24 +192,43 @@ static Expression fromInternal(Expression e, Class<?> targetType) {
     return e;
   }
 
+  /**
+   * build expression list.
+   * if varArgs is true, the last targetType will replace by its Component Type.
+   */
   static List<Expression> fromInternal(Class<?>[] targetTypes,
-      List<Expression> expressions) {
+      List<Expression> expressions, boolean varArgs) {
+
     final List<Expression> list = new ArrayList<>();
-    if (targetTypes.length == expressions.size()) {
+
+    if (varArgs) {
+      Class<?> varTargetType = targetTypes[targetTypes.length - 1];
+      Preconditions.checkArgument(varTargetType != null && varTargetType.isArray(),
+          "the var target type should not null and it must be an array.");
+
```

Comment:
T

Reply:
i think the ? is better in this place,  the question mark represents an unknown type.